### ISSUE TITLE
FCOS_R_50_FPN_1x mAP 36.59

### ISSUE THREAD

**Issue by WitLes** [tianzhi0549/FCOS#102]
Hey, I have run your baseline config "fcos_R_50_FPN_1x.yaml" without any change, but only got 36.59 mAP as reported in 4.1.2 in the paper.
But I noticed that you report 37.1 mAP of the FCOS_R_50_FPN_1x baseline in README.md. Is there any small details I have not noticed?
Thanks!

**Comment by tianzhi0549** [comment]
@WitLes Are you using the latest code and models?

**Comment by WitLes** [comment]
yes. I downloaded your codes several days ago and ran your fcos_R_50_FPN_1x.yaml setting with imagenet pretrained models. I trained the model on coco_2017_train(115k) for two times and got the same mAP 36.59 on coco_2017_val(5k). 
I also downloaded your model in the README.md and tested it, getting 37.06 mAP on coco_2017_val. So it is not caused by the evaluation scripts.

**Comment by tianzhi0549** [comment]
@WitLes It might be because of the multi-GPU training. Please try to train the model with 4 GPUs instead of 8 GPUs.

**Comment by WitLes** [comment]
4GPU, 4 images per GPU? I will try this setting and report it here later.

**Comment by tianzhi0549** [comment]
@WitLes Yes.

**Comment by WitLes** [comment]
I have run 4gpus(GeForce RTX 2080 Ti) config and got 36.809 mAP finally.
I wonder if the drop of mAP is caused by hardware or  packages version.
My config is:
`
MODEL:
  META_ARCHITECTURE: "GeneralizedRCNN"
  WEIGHT: "catalog://ImageNetPretrained/MSRA/R-50"
  RPN_ONLY: True
  FCOS_ON: True
  BACKBONE:
    CONV_BODY: "R-50-FPN-RETINANET"
  RESNETS:
    BACKBONE_OUT_CHANNELS: 256
  RETINANET:
    USE_C5: False # FCOS uses P5 instead of C5
DATASETS:
  TRAIN: ("coco_2017_train", )
  TEST: ("coco_2017_val",)
INPUT:
  MIN_SIZE_TRAIN: (800,)
  MAX_SIZE_TRAIN: 1333
  MIN_SIZE_TEST: 800
  MAX_SIZE_TEST: 1333
DATALOADER:
  SIZE_DIVISIBILITY: 32
SOLVER:
  BASE_LR: 0.01
  WEIGHT_DECAY: 0.0001
  STEPS: (60000, 80000)
  MAX_ITER: 90000
  IMS_PER_BATCH: 16
  WARMUP_METHOD: "constant"
OUTPUT_DIR:
  '/data/fcos_outputs/baseline_4gpu'
`

final mAP:
`
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.368
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.554
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.396
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.208
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.407
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.486
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.313
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.513
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.549
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.334
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.597
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.719
`

**Comment by tianzhi0549** [comment]
@WitLes I am not sure. But do you use python3? We used python2.

**Comment by WitLes** [comment]
Yes..... I use python3.7.3 with pytorch 1.1.0.
Maybe that's the problem.
Still thanks. I will try other configs to see if different models meet the same issue.

**Comment by WitLes** [comment]
I have run "fcos_R_50_FPN_1x.yaml" in [Anaconda2, Python2.7, 4 * 2080ti GPU(4 imgs/gpu), cuda9.0] and finally got 36.9 mAP.
Is this deviation acceptable in your view？

`
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.369
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.555
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.397
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.211
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.410
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.484
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.312
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.515
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.549
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.340
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.602
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.706
`

**Comment by tianzhi0549** [comment]
@WitLes I think it is acceptable. But multiple runs yield the same mAP on my side. The difference might be due to the different versions of packages.

**Comment by WitLes** [comment]
OK. Thank you so much for your reply and advice.

**Comment by haibochina** [comment]
> 好。非常感谢您的答复和建议。

Hi, Have you achieved the high accuracy as the paper of FCOS now?

**Comment by Zzh-tju** [comment]
High AP on pretrained weight ```FCOS_imprv_R_50_FPN_1x.pth```
with default test command
```
python tools/test_net.py --config-file configs/fcos/fcos_imprv_R_50_FPN_1x.yaml MODEL.WEIGHT FCOS_imprv_R_50_FPN_1x.pth TEST.IMS_PER_BATCH 4
```
```
2020-03-30 01:13:16,982 fcos_core.inference INFO: Start evaluation on coco_2014_minival dataset(40504 images).
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10127/10127 [1:28:38<00:00,  1.90it/s]
2020-03-30 02:41:55,565 fcos_core.inference INFO: Total run time: 1:28:38.583243 (0.13131007414996188 s / img per device, on 1 devices)
2020-03-30 02:41:55,565 fcos_core.inference INFO: Model inference time: 1:25:04.233812 (0.12601801826622844 s / img per device, on 1 devices)
2020-03-30 02:41:59,173 fcos_core.inference INFO: Preparing results for COCO format
2020-03-30 02:41:59,174 fcos_core.inference INFO: Preparing bbox results
2020-03-30 02:42:12,525 fcos_core.inference INFO: Evaluating predictions
Loading and preparing results...
DONE (t=22.34s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=439.44s).
Accumulating evaluation results...
DONE (t=114.19s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.475
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.676
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.525
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.295
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.526
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.585
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.368
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.603
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.641
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.442
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.700
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.788
Maximum f-measures for classes:
[0.7942772691314719, 0.6451582707998563, 0.6993149272539809, 0.7297186309633386, 0.8172669793229157, 0.8126338329764454, 0.8532792715285379, 0.5809120703962175, 0.5964369820410952, 0.6528916716581274, 0.882290934645116, 0.8046245063636464, 0.6404191283023074, 0.4846912447680752, 0.6913470882013966, 0.8917793578708526, 0.850948887648513, 0.8188277316905874, 0.7960537506378637, 0.8134604162614276, 0.8847021623491049, 0.9094823672844711, 0.8913231700668929, 0.8996121228058642, 0.45718805615977853, 0.6548161806493983, 0.40931030784082384, 0.6499854550336772, 0.64394488759971, 0.871905065572407, 0.5418841660452399, 0.5867811845831794, 0.7126436781609196, 0.6865767541670825, 0.6588021149345584, 0.7176947599740011, 0.801125703564728, 0.6527583111545064, 0.8158350229958953, 0.6508956792319361, 0.6872131147540983, 0.6595922202294586, 0.5349827897624646, 0.46450959048083773, 0.42146920603512245, 0.6210659086202677, 0.5578648836466766, 0.4532742353912912, 0.580753393320116, 0.5361442014673948, 0.5743179593269179, 0.5310372446936323, 0.6351525900792351, 0.7446323183315834, 0.7559854329909, 0.6259946949602122, 0.5586809596868058, 0.59997416379199, 0.5862137826932227, 0.7041942604856511, 0.5281471621899655, 0.8379202095590529, 0.7731240569164164, 0.778296402392788, 0.7765805506842876, 0.6327546145210557, 0.6995681677976558, 0.6326283660532067, 0.783419689119171, 0.6166822867853796, 0.5486381322957198, 0.6477755738169453, 0.7108462229790454, 0.4483031959041695, 0.8185464999665708, 0.6683242475252111, 0.6090233221305632, 0.7488439332309709, 0.3859269282814614, 0.4982171342779393]
Score thresholds for classes (used in demos for visualization purposes):
[0.49434909224510193, 0.5032018423080444, 0.5023657083511353, 0.5122630596160889, 0.514761745929718, 0.5452229976654053, 0.5591460466384888, 0.5094320178031921, 0.48978444933891296, 0.47464418411254883, 0.5871489644050598, 0.5679255127906799, 0.5206273198127747, 0.46876850724220276, 0.44988346099853516, 0.5506858229637146, 0.5525587201118469, 0.5359309911727905, 0.4905710220336914, 0.5141786336898804, 0.5499591827392578, 0.5743536949157715, 0.5273160934448242, 0.5252215266227722, 0.48926976323127747, 0.48363587260246277, 0.48150262236595154, 0.50196373462677, 0.4925016462802887, 0.5291284322738647, 0.4621402323246002, 0.5136808156967163, 0.474124938249588, 0.47580686211586, 0.5111627578735352, 0.5255421996116638, 0.5037378072738647, 0.49032625555992126, 0.5055979490280151, 0.48306024074554443, 0.5043128728866577, 0.5068569779396057, 0.49172917008399963, 0.45693668723106384, 0.46825405955314636, 0.517708957195282, 0.48183438181877136, 0.4621422588825226, 0.52191561460495, 0.4730696976184845, 0.4853664040565491, 0.4794557988643646, 0.5128921270370483, 0.5329099893569946, 0.4911471903324127, 0.5016376972198486, 0.4824966788291931, 0.48814648389816284, 0.4955405294895172, 0.5073071718215942, 0.48580029606819153, 0.5395932197570801, 0.5383527874946594, 0.5220983624458313, 0.5207992792129517, 0.5123713612556458, 0.5408283472061157, 0.51224684715271, 0.5680729150772095, 0.5118270516395569, 0.4789874255657196, 0.5112487077713013, 0.5208101272583008, 0.4333770275115967, 0.5317476391792297, 0.5033973455429077, 0.5145377516746521, 0.5113539099693298, 0.4739951193332672, 0.4944462478160858]
2020-03-30 02:52:42,582 fcos_core.inference INFO: OrderedDict([('bbox', OrderedDict([('AP', 0.4753139211788734), ('AP50', 0.6763905422157421), ('AP75', 0.5248568894572823), ('APs', 0.29523452437239794), ('APm', 0.5255412113956971), ('APl', 0.5848547417006822)]))])
```

Why does this happen?
